File: dotnet/GoodMem.SemanticKernel/GoodMemClient.cs
using System.Net.Http.Json;
using System.Runtime.CompilerServices;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Web;

namespace GoodMem.SemanticKernel;

/// <summary>
/// Async HTTP client wrapping the GoodMem REST API.
/// This is an internal class; consumers should use <see cref="GoodMemCollection{TRecord}"/>
/// or <see cref="GoodMemVectorStore"/> instead.
/// </summary>
internal sealed class GoodMemClient : IDisposable
{
    private static readonly JsonSerializerOptions s_jsonOptions = new(JsonSerializerDefaults.Web);

    private readonly HttpClient _http;
    private readonly bool _ownsClient;

    internal GoodMemClient(string baseUrl, string apiKey, bool verifySsl = true)
        : this(CreateHttpClient(baseUrl, apiKey, verifySsl), ownsClient: true)
    {
    }

    internal GoodMemClient(HttpClient httpClient, bool ownsClient = false)
    {
        _http = httpClient;
        _ownsClient = ownsClient;
    }

    private static HttpClient CreateHttpClient(string baseUrl, string apiKey, bool verifySsl)
    {
        HttpMessageHandler handler = verifySsl
            ? new HttpClientHandler()
            : new HttpClientHandler { ServerCertificateCustomValidationCallback = (_, _, _, _) => true };

        var client = new HttpClient(handler, disposeHandler: true)
        {
            BaseAddress = new Uri(baseUrl.TrimEnd('/') + "/"),
            Timeout = TimeSpan.FromSeconds(30),
        };
        client.DefaultRequestHeaders.Add("x-api-key", apiKey.Trim());
        return client;
    }

    public void Dispose()
    {
        if (_ownsClient)
            _http.Dispose();
    }

    // ------------------------------------------------------------------
    // Spaces
    // ------------------------------------------------------------------

    internal async Task<List<JsonObject>> ListSpacesAsync(
        string? nameFilter = null,
        CancellationToken ct = default)
    {
        var all = new List<JsonObject>();
        string? nextToken = null;

        do
        {
            var query = HttpUtility.ParseQueryString(string.Empty);
            query["maxResults"] = "1000";
            if (nextToken is not null) query["nextToken"] = nextToken;
            if (nameFilter is not null) query["nameFilter"] = nameFilter;

            var response = await _http.GetAsync($"v1/spaces?{query}", ct).ConfigureAwait(false);
            response.EnsureSuccessStatusCode();

            var data = await response.Content.ReadFromJsonAsync<JsonObject>(s_jsonOptions, ct).ConfigureAwait(false)
                       ?? new JsonObject();

            foreach (var item in data["spaces"]?.AsArray() ?? [])
                if (item is JsonObject obj) all.Add(obj);

            nextToken = data["nextToken"]?.GetValue<string>();
        }
        while (nextToken is not null);

        return all;
    }

    internal async Task<JsonObject> CreateSpaceAsync(
        string name,
        string embedderId,
        string? spaceId = null,
        CancellationToken ct = default)
    {
        var payload = new JsonObject
        {
            ["name"] = name,
            ["spaceEmbedders"] = new JsonArray(new JsonObject
            {
                ["embedderId"] = embedderId,
                ["defaultRetrievalWeight"] = 1.0,
            }),
            ["defaultChunkingConfig"] = new JsonObject
            {
                ["recursive"] = new JsonObject
                {
                    ["chunkSize"] = 512,
                    ["chunkOverlap"] = 64,
                    ["keepStrategy"] = "KEEP_END",
                    ["lengthMeasurement"] = "CHARACTER_COUNT",
                },
            },
        };
        if (spaceId is not null)
            payload["spaceId"] = spaceId;

        var response = await _http.PostAsJsonAsync("v1/spaces", payload, s_jsonOptions, ct).ConfigureAwait(false);
        response.EnsureSuccessStatusCode();
        return await response.Content.ReadFromJsonAsync<JsonObject>(s_jsonOptions, ct).ConfigureAwait(false)
               ?? new JsonObject();
    }

    internal async Task DeleteSpaceAsync(string spaceId, CancellationToken ct = default)
    {
        var response = await _http.DeleteAsync($"v1/spaces/{Uri.EscapeDataString(spaceId)}", ct).ConfigureAwait(false);
        response.EnsureSuccessStatusCode();
    }

    // ------------------------------------------------------------------
    // Embedders
    // ------------------------------------------------------------------

    internal async Task<List<JsonObject>> ListEmbeddersAsync(CancellationToken ct = default)
    {
        var response = await _http.GetAsync("v1/embedders", ct).ConfigureAwait(false);
        response.EnsureSuccessStatusCode();

        var data = await response.Content.ReadFromJsonAsync<JsonObject>(s_jsonOptions, ct).ConfigureAwait(false)
                   ?? new JsonObject();

        var result = new List<JsonObject>();
        foreach (var item in data["embedders"]?.AsArray() ?? [])
            if (item is JsonObject obj) result.Add(obj);
        return result;
    }

    // ------------------------------------------------------------------
    // Memories
    // ------------------------------------------------------------------

    internal async Task<JsonObject> CreateMemoryAsync(
        string spaceId,
        string content,
        string contentType = "text/plain",
        Dictionary<string, object?>? metadata = null,
        string? memoryId = null,
        CancellationToken ct = default)
    {
        var payload = new JsonObject
        {
            ["spaceId"] = spaceId,
            ["originalContent"] = content,
            ["contentType"] = contentType,
        };
        if (metadata is { Count: > 0 })
            payload["metadata"] = JsonSerializer.SerializeToNode(metadata, s_jsonOptions);
        if (memoryId is not null)
            payload["memoryId"] = memoryId;

        var response = await _http.PostAsJsonAsync("v1/memories", payload, s_jsonOptions, ct).ConfigureAwait(false);
        response.EnsureSuccessStatusCode();
        return await response.Content.ReadFromJsonAsync<JsonObject>(s_jsonOptions, ct).ConfigureAwait(false)
               ?? new JsonObject();
    }

    internal async Task<List<JsonObject>> BatchGetMemoriesAsync(
        IEnumerable<string> memoryIds,
        CancellationToken ct = default)
    {
        var ids = memoryIds.ToList();
        if (ids.Count == 0) return [];

        var payload = new JsonObject
        {
            ["memoryIds"] = new JsonArray(ids.Select(id => JsonValue.Create(id)).Cast<JsonNode?>().ToArray()),
            ["includeContent"] = true,
        };

        var response = await _http.PostAsJsonAsync("v1/memories:batchGet", payload, s_jsonOptions, ct).ConfigureAwait(false);
        response.EnsureSuccessStatusCode();

        var data = await response.Content.ReadFromJsonAsync<JsonObject>(s_jsonOptions, ct).ConfigureAwait(false)
                   ?? new JsonObject();

        // Response shape: { "results": [ { "success": true, "memory": {...} }, ... ] }
        var result = new List<JsonObject>();
        foreach (var item in data["results"]?.AsArray() ?? [])
        {
            if (item is JsonObject resultObj &&
                resultObj["success"]?.GetValue<bool>() == true &&
                resultObj["memory"] is JsonObject mem)
            {
                result.Add(mem);
            }
        }
        return result;
    }

    internal async Task DeleteMemoryAsync(string memoryId, CancellationToken ct = default)
    {
        var response = await _http.DeleteAsync(
            $"v1/memories/{Uri.EscapeDataString(memoryId)}", ct).ConfigureAwait(false);

        if (response.StatusCode == System.Net.HttpStatusCode.NotFound)
            return;

        response.EnsureSuccessStatusCode();
    }

    /// <summary>
    /// Semantic search over one or more spaces. Parses the NDJSON response and
    /// returns a unified list of results with chunk, memory, and score.
    /// </summary>
    internal async Task<List<RetrieveResult>> RetrieveMemoriesAsync(
        string query,
        IEnumerable<string> spaceIds,
        int top = 5,
        CancellationToken ct = default)
    {
        var payload = new JsonObject
        {
            ["message"] = query,
            ["spaceKeys"] = new JsonArray(
                spaceIds.Select(sid => (JsonNode?)new JsonObject { ["spaceId"] = sid }).ToArray()),
            ["requestedSize"] = top,
        };

        using var request = new HttpRequestMessage(HttpMethod.Post, "v1/memories:retrieve")
        {
            Content = JsonContent.Create(payload, options: s_jsonOptions),
        };
        request.Headers.Accept.ParseAdd("application/x-ndjson");

        using var response = await _http.SendAsync(request, ct).ConfigureAwait(false);
        response.EnsureSuccessStatusCode();

        var body = await response.Content.ReadAsStringAsync(ct).ConfigureAwait(false);

        // Parse NDJSON events.
        // memoryDefinition events arrive indexed by position (0, 1, 2, …).
        // retrievedItem events reference a memoryIndex into that list.
        var memoryList = new List<JsonObject>();
        var chunkRefs = new List<JsonObject>();

        foreach (var line in body.Split('\n', StringSplitOptions.RemoveEmptyEntries))
        {
            var trimmed = line.Trim();
            if (string.IsNullOrEmpty(trimmed)) continue;

            JsonObject? evt;
            try { evt = JsonSerializer.Deserialize<JsonObject>(trimmed, s_jsonOptions); }
            catch { continue; }
            if (evt is null) continue;

            if (evt["memoryDefinition"] is JsonObject memDef)
            {
                memoryList.Add(memDef);
                continue;
            }

            if (evt["retrievedItem"] is JsonObject item && item["chunk"] is JsonObject chunkRef)
                chunkRefs.Add(chunkRef);
        }

        var results = new List<RetrieveResult>(chunkRefs.Count);
        foreach (var chunkRef in chunkRefs)
        {
            var chunk = chunkRef["chunk"] as JsonObject ?? new JsonObject();
            var rawScore = chunkRef["relevanceScore"]?.GetValue<double>() ?? 0.0;
            var memoryIndex = chunkRef["memoryIndex"]?.GetValue<int>();

            var mem = memoryIndex is >= 0 && memoryIndex < memoryList.Count
                ? memoryList[memoryIndex.Value]
                : new JsonObject();

            results.Add(new RetrieveResult(chunk, mem, -rawScore)); // negate: lower-is-better → higher-is-better
        }

        return results;
    }

    /// <summary>A correlated retrieve result: chunk data, memory metadata, and similarity score.</summary>
    internal sealed record RetrieveResult(JsonObject Chunk, JsonObject Memory, double Score);
}

File: java/goodmem-semantic-kernel/src/main/java/ai/goodmem/semantickernel/GoodMemClient.java
package ai.goodmem.semantickernel;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import reactor.core.publisher.Mono;
import reactor.core.scheduler.Schedulers;

import javax.net.ssl.SSLContext;
import javax.net.ssl.TrustManager;
import javax.net.ssl.X509TrustManager;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.security.SecureRandom;
import java.security.cert.X509Certificate;
import java.time.Duration;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Internal HTTP client wrapping the GoodMem REST API.
 *
 * <p>All methods are package-private. Consumers should use
 * {@link GoodMemCollection} or {@link GoodMemVectorStore} instead.
 *
 * <p>Each method wraps a blocking {@link HttpClient#send} call inside
 * {@code Mono.fromCallable(...).subscribeOn(Schedulers.boundedElastic())}
 * so it integrates cleanly with Project Reactor without tying up the
 * event-loop thread.
 */
final class GoodMemClient implements AutoCloseable {

    static final ObjectMapper MAPPER = new ObjectMapper();
    private static final Duration TIMEOUT = Duration.ofSeconds(30);

    private final HttpClient http;
    private final String baseUrl;
    private final String apiKey;

    GoodMemClient(GoodMemOptions options) {
        this.baseUrl = options.getBaseUrl().replaceAll("/+$", "");
        this.apiKey  = options.getApiKey().strip();
        this.http    = options.isVerifySsl() ? defaultClient() : insecureClient();
    }

    /** Package-private constructor for tests — injects a pre-built HttpClient. */
    GoodMemClient(HttpClient http, String baseUrl, String apiKey) {
        this.http    = http;
        this.baseUrl = baseUrl.replaceAll("/+$", "");
        this.apiKey  = apiKey;
    }

    @Override
    public void close() {
        // java.net.http.HttpClient does not implement Closeable before Java 21.
        // Nothing to release here; the JVM will GC the underlying connections.
    }

    // ── Spaces ────────────────────────────────────────────────────────────────

    Mono<List<ObjectNode>> listSpaces(String nameFilter) {
        return Mono.fromCallable(() -> {
            List<ObjectNode> all = new ArrayList<>();
            String nextToken = null;
            do {
                StringBuilder qs = new StringBuilder("?maxResults=1000");
                if (nextToken  != null) qs.append("&nextToken=").append(encode(nextToken));
                if (nameFilter != null) qs.append("&nameFilter=").append(encode(nameFilter));

                ObjectNode data = doGet("v1/spaces" + qs);
                for (JsonNode item : iterArray(data, "spaces"))
                    if (item.isObject()) all.add((ObjectNode) item);

                String tok = data.path("nextToken").asText(null);
                nextToken = (tok != null && !tok.isEmpty()) ? tok : null;
            } while (nextToken != null);
            return all;
        }).subscribeOn(Schedulers.boundedElastic());
    }

    /**
     * @param spaceId optional — when non-null it is sent as the requested UUID;
     *                the server may honour or ignore it.
     */
    Mono<ObjectNode> createSpace(String name, String embedderId, String spaceId) {
        return Mono.fromCallable(() -> {
            ObjectNode payload = MAPPER.createObjectNode().put("name", name);

            ArrayNode embedders = payload.putArray("spaceEmbedders");
            embedders.addObject()
                    .put("embedderId", embedderId)
                    .put("defaultRetrievalWeight", 1.0);

            payload.putObject("defaultChunkingConfig")
                    .putObject("recursive")
                    .put("chunkSize", 512)
                    .put("chunkOverlap", 64)
                    .put("keepStrategy", "KEEP_END")
                    .put("lengthMeasurement", "CHARACTER_COUNT");

            if (spaceId != null) payload.put("spaceId", spaceId);
            return doPost("v1/spaces", payload);
        }).subscribeOn(Schedulers.boundedElastic());
    }

    Mono<Void> deleteSpace(String spaceId) {
        return Mono.fromCallable(() -> {
            doDelete("v1/spaces/" + encode(spaceId), false);
            return (Void) null;
        }).subscribeOn(Schedulers.boundedElastic()).then();
    }

    // ── Embedders ─────────────────────────────────────────────────────────────

    Mono<List<ObjectNode>> listEmbedders() {
        return Mono.fromCallable(() -> {
            ObjectNode data = doGet("v1/embedders");
            List<ObjectNode> result = new ArrayList<>();
            for (JsonNode item : iterArray(data, "embedders"))
                if (item.isObject()) result.add((ObjectNode) item);
            return result;
        }).subscribeOn(Schedulers.boundedElastic());
    }

    // ── Memories ──────────────────────────────────────────────────────────────

    Mono<ObjectNode> createMemory(
            String spaceId,
            String content,
            String contentType,
            Map<String, Object> metadata,
            String memoryId) {
        return Mono.fromCallable(() -> {
            ObjectNode payload = MAPPER.createObjectNode()
                    .put("spaceId", spaceId)
                    .put("originalContent", content)
                    .put("contentType", contentType != null ? contentType : "text/plain");

            if (metadata != null && !metadata.isEmpty())
                payload.set("metadata", MAPPER.valueToTree(metadata));
            if (memoryId != null)
                payload.put("memoryId", memoryId);

            return doPost("v1/memories", payload);
        }).subscribeOn(Schedulers.boundedElastic());
    }

    Mono<List<ObjectNode>> batchGetMemories(List<String> memoryIds) {
        if (memoryIds.isEmpty()) return Mono.just(List.of());

        return Mono.fromCallable(() -> {
            ObjectNode payload = MAPPER.createObjectNode();
            ArrayNode ids = payload.putArray("memoryIds");
            memoryIds.forEach(ids::add);
            payload.put("includeContent", true);

            ObjectNode data = doPost("v1/memories:batchGet", payload);

            // Response: { "results": [ { "success": true, "memory": {...} }, ... ] }
            List<ObjectNode> result = new ArrayList<>();
            for (JsonNode item : iterArray(data, "results")) {
                if (item.path("success").asBoolean() && item.path("memory").isObject())
                    result.add((ObjectNode) item.get("memory"));
            }
            return result;
        }).subscribeOn(Schedulers.boundedElastic());
    }

    /** 404 responses are silently ignored (already deleted). */
    Mono<Void> deleteMemory(String memoryId) {
        return Mono.fromCallable(() -> {
            doDelete("v1/memories/" + encode(memoryId), true);
            return (Void) null;
        }).subscribeOn(Schedulers.boundedElastic()).then();
    }

    /**
     * Semantic search over one or more spaces.
     *
     * <p>GoodMem returns a streaming NDJSON body. Each line is a JSON event:
     * <ul>
     *   <li>{@code {"memoryDefinition": {...}}} — arrives indexed by position 0, 1, 2, …
     *   <li>{@code {"retrievedItem": {"chunk": {"memoryIndex": N, "relevanceScore": …}}}}
     *       — references a memory definition by index
     * </ul>
     *
     * <p>Scores are negated before returning: GoodMem uses lower-is-better (distance),
     * SK uses higher-is-better (similarity).
     */
    Mono<List<RetrieveResult>> retrieveMemories(String query, List<String> spaceIds, int top) {
        return Mono.fromCallable(() -> {
            ObjectNode payload = MAPPER.createObjectNode()
                    .put("message", query)
                    .put("requestedSize", top);

            ArrayNode keys = payload.putArray("spaceKeys");
            spaceIds.forEach(sid -> keys.addObject().put("spaceId", sid));

            String json = MAPPER.writeValueAsString(payload);
            HttpRequest request = baseRequest("v1/memories:retrieve")
                    .header("Accept", "application/x-ndjson")
                    .POST(HttpRequest.BodyPublishers.ofString(json))
                    .build();

            HttpResponse<String> response = http.send(request, HttpResponse.BodyHandlers.ofString());
            ensureSuccess(response);

            List<ObjectNode> memoryList = new ArrayList<>();
            List<ObjectNode> chunkRefs  = new ArrayList<>();

            for (String line : response.body().split("\n")) {
                String trimmed = line.strip();
                if (trimmed.isEmpty()) continue;

                JsonNode evt;
                try { evt = MAPPER.readTree(trimmed); } catch (Exception e) { continue; }

                JsonNode memDef = evt.path("memoryDefinition");
                if (memDef.isObject()) {
                    memoryList.add((ObjectNode) memDef);
                    continue;
                }

                JsonNode chunkRef = evt.path("retrievedItem").path("chunk");
                if (chunkRef.isObject()) chunkRefs.add((ObjectNode) chunkRef);
            }

            List<RetrieveResult> results = new ArrayList<>(chunkRefs.size());
            for (ObjectNode chunkRef : chunkRefs) {
                JsonNode chunkNode = chunkRef.path("chunk");
                ObjectNode chunk = chunkNode.isObject()
                        ? (ObjectNode) chunkNode
                        : MAPPER.createObjectNode();

                double rawScore  = chunkRef.path("relevanceScore").asDouble(0.0);
                int    memIndex  = chunkRef.path("memoryIndex").asInt(-1);
                ObjectNode mem   = (memIndex >= 0 && memIndex < memoryList.size())
                        ? memoryList.get(memIndex)
                        : MAPPER.createObjectNode();

                results.add(new RetrieveResult(chunk, mem, -rawScore)); // negate → higher-is-better
            }
            return results;
        }).subscribeOn(Schedulers.boundedElastic());
    }

    // ── Internal HTTP helpers ─────────────────────────────────────────────────

    private ObjectNode doGet(String path) throws Exception {
        HttpResponse<String> resp = http.send(
                baseRequest(path).GET().build(),
                HttpResponse.BodyHandlers.ofString());
        ensureSuccess(resp);
        return (ObjectNode) MAPPER.readTree(resp.body());
    }

    private ObjectNode doPost(String path, ObjectNode body) throws Exception {
        String json = MAPPER.writeValueAsString(body);
        HttpResponse<String> resp = http.send(
                baseRequest(path)
                        .POST(HttpRequest.BodyPublishers.ofString(json))
                        .build(),
                HttpResponse.BodyHandlers.ofString());
        ensureSuccess(resp);
        return (ObjectNode) MAPPER.readTree(resp.body());
    }

    private void doDelete(String path, boolean notFoundOk) throws Exception {
        HttpResponse<String> resp = http.send(
                baseRequest(path).DELETE().build(),
                HttpResponse.BodyHandlers.ofString());
        if (notFoundOk && resp.statusCode() == 404) return;
        ensureSuccess(resp);
    }

    private HttpRequest.Builder baseRequest(String path) {
        return HttpRequest.newBuilder()
                .uri(URI.create(baseUrl + "/" + path))
                .header("x-api-key", apiKey)
                .header("Content-Type", "application/json")
                .header("Accept", "application/json")
                .timeout(TIMEOUT);
    }

    private static void ensureSuccess(HttpResponse<String> response) {
        if (response.statusCode() >= 400)
            throw new GoodMemException(
                    "GoodMem API error " + response.statusCode() + ": " + response.body());
    }

    private static String encode(String value) {
        return URLEncoder.encode(value, StandardCharsets.UTF_8);
    }

    /** Null-safe iteration over a named JSON array field. */
    private static Iterable<JsonNode> iterArray(JsonNode parent, String field) {
        JsonNode arr = parent.path(field);
        return arr.isArray() ? arr : List.of();
    }

    private static HttpClient defaultClient() {
        return HttpClient.newBuilder()
                .connectTimeout(Duration.ofSeconds(10))
                .build();
    }

    @SuppressWarnings("java:S4830") // intentional: SSL verification disabled by user request
    private static HttpClient insecureClient() {
        try {
            TrustManager[] trustAll = {new X509TrustManager() {
                public void checkClientTrusted(X509Certificate[] c, String a) {}
                public void checkServerTrusted(X509Certificate[] c, String a) {}
                public X509Certificate[] getAcceptedIssuers() { return new X509Certificate[0]; }
            }};
            SSLContext ctx = SSLContext.getInstance("TLS");
            ctx.init(null, trustAll, new SecureRandom());
            return HttpClient.newBuilder()
                    .sslContext(ctx)
                    .connectTimeout(Duration.ofSeconds(10))
                    .build();
        } catch (Exception e) {
            throw new GoodMemException("Failed to build insecure SSL context", e);
        }
    }

    /** A correlated retrieve result: chunk data, parent memory metadata, and similarity score. */
    record RetrieveResult(ObjectNode chunk, ObjectNode memory, double score) {}
}
